### DISCUSSION TITLE
Is it possible to prevent encoding when setting search parameters?

### DISCUSSION THREAD

**Discussion by MA-MacDonald** [remix-run/react-router#9752] [Q&A]
Is there a way to prevent encoding when setting search parameters? 

For example when I do:
`setSearchParams({ filters: "foo[bar][baz]=foobarbaz" });`

I get:
`?filters=foo%5Bbar%5D%5Bbaz%5D%3Dfoobarbaz`

I want it to put the exact string in the url like this:
`?filters=foo[bar][baz]=foobarbaz`

**Comment by kiliman** [comment]
No. That's just how the native `URLSearchParams` works. You really shouldn't care what the URL looks like in the address bar. 

Browsers tend to be forgiving in what you can enter as the URL, but the values are encoded to ensure the parser will not treat any embedded `=` or (`?`, `&`, etc) as part of the querystring syntax. 

Perhaps it would have been nice if it _only_ encoded those characters instead of the entire thing.

**Reply by lancejpollard** [reply]
You should care what the URL likes like in the address bar.

**Reply by GabenGar** [reply]
It's also not the library's job to operate on "readable" URLs, especially since they can be used for scamming.
The OP assumed you can shove any values into search params unchanged, presumably for PHP backend (which has a hardon for madeup serialization formats using search params) so it's a design problem of the application first and foremost.

**Comment by smeijer** [comment]
Ironically, it's exactly the "native encoding" that causes issues with Vite.

<img width="303" alt="image" src="https://github.com/user-attachments/assets/d006d039-6caa-4d43-ba5c-ab1ff8c5324b" />
